### PR TITLE
[WIP] Look for string.GetPinnableReference when pinning a string first

### PR DESCRIPTION
**Pull request by cartermp**
This tries to address #9251 

I couldn't actually run these tests on my machine with VS for some weird reason, so I'll let CI tell me what I did wrong instead

### CODE REVIEW

File: src/fsharp/TypeChecker.fs [hunk 1]
```diff
@@ -3099,9 +3099,16 @@ let BuildDisposableCleanup cenv env m (v: Val) =
 let BuildOffsetToStringData cenv env m = 
     let ad = env.eAccessRights
     let offsetToStringDataMethod = 
-        match TryFindIntrinsicOrExtensionMethInfo ResultCollectionSettings.AllResults cenv env m ad "get_OffsetToStringData" cenv.g.system_RuntimeHelpers_ty with 
+        // Prefer string.GetPinnableReference introduced in .NET Core 3
```
Comment (KevinRansom) [inline src/fsharp/TypeChecker.fs:3102]: @cartermp , this is code generation.  So the reflection to find the API, is happening on the machine the compiler is executing on.  So probably we need to generate a helper function to do the call, and then to call that.

Comment (cartermp) [reply src/fsharp/TypeChecker.fs:3102]: I don't think I understand. There's no reflection involved here? It's just checking of the type has a given method in its method table that's already been imported.

Comment (KevinRansom) [reply src/fsharp/TypeChecker.fs:3102]: I think you are right, it checks the referenced types rather than the executing types.

**Comment by GrabYourPitchforks** [comment]
Thanks kindly for getting to this so quickly! :)

File: src/fsharp/TypeChecker.fs [hunk 2]
```diff
@@ -11006,10 +11013,18 @@ and TcAndBuildFixedExpr cenv env (overallPatTy, fixedExpr, overallExprTy, mBindi
     | ty when isStringTy cenv.g ty -> 
         let charPtrTy = mkNativePtrTy cenv.g cenv.g.char_ty
         UnifyTypes cenv env mBinding charPtrTy overallPatTy
+        
+        // When targeting .NET Core or higher:
+        //
+        //    let ptr: nativeptr<char> = 
+        //        let pinned s = str
+        //        (nativeptr)s + GetPinnableReference()
```
Comment (GrabYourPitchforks) [inline src/fsharp/TypeChecker.fs:11021]: Just to clarify, the compiler shouldn't perform any arithmetic here. The result of `GetPinnableReference` is a _readonly char\&_, and that _char\&_ itself should be pinned. The string instance itself should not be pinned.

Comment (cartermp) [reply src/fsharp/TypeChecker.fs:11021]: Hmmm. @dsyme wrote this originally, I assume he had a good reason for doing this, but yes it shouldn't need to do that.

Comment (GrabYourPitchforks) [reply src/fsharp/TypeChecker.fs:11021]: Clarification to my comments: the `GetOffsetToStringData` pattern works by having the caller pin the string instance, then compute `addrof(string) + GetOffsetToStringData`, then expose the resulting sum as a _char\*_. So the addition is expected there.

The `GetPinnableReference` pattern works by providing the _char\&_ directly, which the caller then pins.

Comment (cartermp) [reply src/fsharp/TypeChecker.fs:11021]: Gotcha, I'll take another look at this

File: tests/fsharp/Compiler/CodeGen/EmittedIL/FixedExpressionCodeGen.fs [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    let ``Pinning a string emits string.GetPinnableReference`` () =
+        let code = """
+module FixedOnAString
+#nowarn "9"
+let foo(s : string) =
+    use ptr = fixed s
+    ptr
+"""
+        let il = """
+char& modreq([System.Runtime]System.Runtime.InteropServices.InAttribute) [System.Runtime]System.String::GetPinnableReference()
```
Comment (TIHan) [inline tests/fsharp/Compiler/CodeGen/EmittedIL/FixedExpressionCodeGen.fs:20]: We should check to make sure the local char is pinned as well from locals init.

Comment (cartermp) [reply tests/fsharp/Compiler/CodeGen/EmittedIL/FixedExpressionCodeGen.fs:20]: Yeah, thinking about it more this is a bad test. I'll revisit this one

**Comment by cartermp** [comment]
The fix here isn't right and will require a bit more work in the typechecker to do the right thing. I'm going to close this one out, at least until some of the test refactoring is completed, so that I don't have to baby the PR along without a proper fix.

**Pull request closed without merging**